src/addressresolver.js

```diff
@@ -0,0 +1,76 @@
+const {SolanaDomainConverter} = require('./solanadomainconverter.js');
+
+class AddressResolver {
+  constructor(hash, env) {
+    this.hash = hash;
+    this.env = env;
+  }
+
+  async getDestinationURL() {
+    if (!this.attemptedToResolve) {
+      await this.resolve_();
+    }
+    return this.url;
+  }
+
+  async resolve_() {
+    if (this.attemptedToResolve) {
+      return;
+    }
+    this.attemptedToResolve = true;
+    const coinType = getCoinTypeForHash_(this.hash);
+    switch (coinType) {
+      case COIN_TYPE.ETHEREUM:
+        this.resolveEthereumAddr_();
+        return;
+      case COIN_TYPE.SOLANA:
+      default:
+        await this.resolveSolanaAddr_();
+        return;
+    }
+  }
+
+  resolveEthereumAddr_() {
+    this.url = etherscanURL_(`address/${this.hash}`, this.env);
+  }
+
+  async resolveSolanaAddr_() {
+    let address = this.hash;
+    if (this.hash.includes('.sol')) {
+      address = await (new SolanaDomainConverter(this.hash, this.env))
+        .toWalletAddress();
+    }
+    this.url = solScanURL_(`account/${address}`, this.env);
+  }
+}
+
+function getCoinTypeForHash_(hash) {
+  if (hash.startsWith('0x') || hash.includes('.eth')) {
+    return COIN_TYPE.ETHEREUM;
+  }
+  return COIN_TYPE.SOLANA;
+}
+
+function solScanURL_(path, env) {
```

Comment:
is the _ postfix to indicate it's private? I have seen the prefix version of it, `_privateFunction()`, never seen this one

Reply:
It's a google thing https://google.github.io/styleguide/jsguide.html#jsdoc-visibility-annotations we can do either way